source/extensions/http/header_validators/envoy_default/http2_header_validator.cc

```diff
@@ -97,6 +106,104 @@ Http2HeaderValidator::validateResponseHeaderEntry(const HeaderString& key,
   return validateGenericHeaderValue(value);
 }
 
+void ServerHttp2HeaderValidator::encodeExtendedAsciiInPath(
+    ::Envoy::Http::RequestHeaderMap& header_map) {
+  absl::string_view path = header_map.path();
+  // Check if URI the path contains any characters >= 0x80
+  auto extended_ascii_position = path.begin();
+  for (; extended_ascii_position != path.end() &&
+         static_cast<unsigned char>(*extended_ascii_position) < 0x7F;
+       ++extended_ascii_position) {
+    // Return early if we got to query or fragment without finding any extended ASCII characters.
+    if (*extended_ascii_position == '?' || *extended_ascii_position == '#') {
+      return;
+    }
+  }
+  if (extended_ascii_position == path.end()) {
+    return;
+  }
+  std::string encoded_path(path.begin(), extended_ascii_position);
+  encoded_path.reserve(path.size());
+
+  for (; extended_ascii_position != path.end(); ++extended_ascii_position) {
+    if (*extended_ascii_position == '?' || *extended_ascii_position == '#') {
+      break;
+    }
+    if (static_cast<unsigned char>(*extended_ascii_position) >= 0x80) {
+      absl::StrAppend(&encoded_path, fmt::format("%{:02X}", static_cast<const unsigned char&>(
+                                                                *extended_ascii_position)));
+    } else {
+      encoded_path.push_back(*extended_ascii_position);
+    }
+  }
+  // Append query and fragment if present
+  encoded_path.append(extended_ascii_position, path.end());
+  // Encoding would change the length of the path
+  if (encoded_path.size() > path.size()) {
+    header_map.setPath(encoded_path);
+  }
+}
+
+HeaderValidator::HeaderValueValidationResult
+Http2HeaderValidator::validatePathHeaderCharactersExtendedAsciiAllowed(const HeaderString& value) {
+  const auto& path = value.getStringView();
+  bool is_valid = !path.empty();
+
+  // When the envoy.reloadable_features.uhv_translate_backslash_to_slash == true
+  // the validation method needs to allow backslashes in path, so they can be translated
+  // to slashes during path normalization.
+  const bool allow_backslash =
+      Runtime::runtimeFeatureEnabled("envoy.reloadable_features.uhv_translate_backslash_to_slash");
+
+  auto iter = path.begin();
+  auto end = path.end();
+  // Validate the path component of the URI
+  for (; iter != end && is_valid; ++iter) {
+    const char ch = *iter;
+    if ((allow_backslash && ch == '\\') || static_cast<unsigned char>(ch) >= 0x80) {
+      continue;
+    }
+    if (ch == '?' || ch == '#') {
+      // This is the start of the query or fragment portion of the path which uses a different
+      // character table.
+      break;
+    }
+
+    is_valid &= testCharInTable(kPathHeaderCharTable, ch);
+  }
+
+  if (is_valid && iter != end && *iter == '?') {
+    // Validate the query component of the URI
+    ++iter;
+    for (; iter != end && is_valid; ++iter) {
+      const char ch = *iter;
```

Comment:
Up to you, but I'd consider eliminating this variable and just using *iter down on line 183 since we're not using it on line 180. (Or vice versa, I suppose)

Reply:
Fixed